File: cflib_2_12/csharp/CFLib/CFLib.cs
/*
 *  MSS Code Factory CFLib 2.12
 *
 *	Copyright (c) 2019 Mark Stephen Sobkow
 *	
 *	Licensed under the Apache License, Version 2.0 (the "License");
 *	you may not use this file except in compliance with the License.
 *	You may obtain a copy of the License at
 *	
 *		http://www.apache.org/licenses/LICENSE-2.0
 *	
 *	Unless required by applicable law or agreed to in writing, software
 *	distributed under the License is distributed on an "AS IS" BASIS,
 *	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *	See the License for the specific language governing permissions and
 *	limitations under the License.
 */

using System;
using System.Globalization;

namespace CFLib
{
    public class CFLib {
        public static char[] S_Empty = { '\0' };
        private static String linkName = "CFLib";
        public static String LinkName
        {
            get
            {
                return (linkName);
            }
        }

        private static String linkVersion = "2.12.11194";
		public static String LinkVersion
        {
            get
            {
                return (linkVersion);
            }
        }

		private static ICFLibExceptionFactory defaultExceptionFactory = (ICFLibExceptionFactory)( new CFLibEnCaExceptionFactory() );
        public static ICFLibExceptionFactory DefaultExceptionFactory
        {
            get
            {
                return (defaultExceptionFactory);
            }

            set {
                if (value == null)
                {
                    throw defaultExceptionFactory.newNullArgumentException(typeof(CFLib),
					    "setExceptionFactory",
					    1,
					    "value",
					    null );
			    }
                defaultExceptionFactory = value;
		    }
        }
 
		public static TimeZoneInfo UTC_TIMEZONE = TimeZoneInfo.Utc;
 
		public static void beep() {
            /*
			Clip clip = null;
			// The audio is courtesy of a whole whack of articles from stackoverflow.com, each of which got me one line closer to working
			try {
				InputStream resource = typeof(CFLib).getResourceAsStream("/net/sourceforge/msscf/cflib/CFLib/sounds/alert.wav");
				if( resource != null ) {
					AudioInputStream audioInputStream = AudioSystem.getAudioInputStream( new BufferedInputStream( resource ) );
					DataLine.Info info = new DataLine.Info( typeof(Clip), audioInputStream.getFormat() );
					clip = (Clip)AudioSystem.getLine( info );
					clip.open( audioInputStream );
					clip.start();
					clip.drain();
				}
			} catch (LineUnavailableException e) {
			} catch (UnsupportedAudioFileException e) {
			} catch (IOException e) {
			} catch (NullPointerException e) {
			} catch (IllegalArgumentException e) {
			}
            */
		}

		public static DateTime getUTCGregorianDateTime(
			int year,
			int month,
			int day,
			int hour,
			int minute,
			int second )
		{
            DateTime cal = new DateTime( year, month, day, hour, minute, second, DateTimeKind.Utc );
			return( cal );
		}
    }
}

File: cflib_2_12/java/cflib/src/org.msscf.msscf.cflib.CFLib/org/msscf/msscf/cflib/CFLib/CFLib.java
/*
 *  MSS Code Factory CFLib 2.12
 *
 *	Copyright 2016-2026 Mark Stephen Sobkow
 *
 *	These files are part of MSS Code Factory Version 2.12.
 *
 *	MSS Code Factory is available under the terms of the GNU General Public License,
 *	Version 3 or later.
 *
 *	MSS Code Factory is free software: you can redistribute it and/or
 *	modify it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation, either version 3 of the License, or
 *	(at your option) any later version.
 *
 *	MSS Code Factory is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 */

package org.msscf.msscf.cflib.CFLib;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.TimeZone;

import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.Clip;
import javax.sound.sampled.DataLine;
import javax.sound.sampled.LineUnavailableException;
import javax.sound.sampled.UnsupportedAudioFileException;

/**
 *	The performance of exception handling in a high-volume Java system cannot
 *	be underestimated.  Anything you can do to speed up exception
 *	instantiation and initialization will dramatically improve the
 *	performance of the system under real life workloads, where users
 *	do input bad data.
 *	<p>
 *	The exception factory model eliminates the resolution of NLS translations
 *	of key system implementation error conditions using a plug-in/registry
 *	model instead of tolerating the resource load overhead of the default
 *	Java NLS string resource approach.
 */
public class CFLib {
	public final static String LinkName = "CFLib";
	public final static String LinkVersion = "2.12.11194";

	public final static TimeZone UTC_TIMEZONE = TimeZone.getTimeZone( "+0000" );

	public static void beep() {
		Clip clip = null;
		// The audio is courtesy of a whole whack of articles from stackoverflow.com, each of which got me one line closer to working
		try {
			InputStream resource = CFLib.class.getResourceAsStream("/org/msscf/msscf/cflib/CFLib/sounds/alert.wav");
			if( resource != null ) {
				AudioInputStream audioInputStream = AudioSystem.getAudioInputStream( new BufferedInputStream( resource ) );
				DataLine.Info info = new DataLine.Info( Clip.class, audioInputStream.getFormat() );
				clip = (Clip)AudioSystem.getLine( info );
				clip.open( audioInputStream );
				clip.start();
				clip.drain();
			}
		} catch (LineUnavailableException e) {
		} catch (UnsupportedAudioFileException e) {
		} catch (IOException e) {
		} catch (NullPointerException e) {
		} catch (IllegalArgumentException e) {
		}
	}

	public static Calendar getUTCGregorianCalendar(
		int year,
		int month,
		int day,
		int hour,
		int minute,
		int second )
	{
		Calendar cal = new GregorianCalendar( UTC_TIMEZONE );
		cal.clear();
		cal.set( year, month, day, hour, minute, second );
        cal.getTimeInMillis(); // Force calendar resync based on input values
		return( cal );
	}
	
	public String getLinkName() {
		return( LinkName );
	}

	public String getLinkVersion() {
		return( LinkVersion );
	}
}
